C++: Improve left shift and right shift range analysis accuracy

gsingh93 opened the pull request:
This PR adds the `ConstantLShiftExprRange` and `ConstantRShiftExprRange` classes, which extend `SimpleRangeAnalysisExpr`, to improve the accuracy of the range analysis. Implementing these improvements in this way was easier than modifying the core range analysis code, and these improvements have led to finding security bugs in production code, so I wanted to contribute this upstream.

To see the before and after of the test, see the diff between the two commits.

gsingh93 commented on cpp/ql/lib/experimental/semmle/code/cpp/rangeanalysis/extensions/ConstantRShiftExprRange.qll:
```diff
@@ -0,0 +1,136 @@
+private import experimental.semmle.code.cpp.models.interfaces.SimpleRangeAnalysisExpr
+private import semmle.code.cpp.rangeanalysis.RangeAnalysisUtils
+
+float evaluateConstantExpr1(Expr e) {
```
This is duplicated in both files, is there some common file I can move it to?

gsingh93 commented on cpp/ql/test/experimental/library-tests/rangeanalysis/bitshift/BitShiftRange.ql:
```diff
@@ -0,0 +1,25 @@
+import experimental.semmle.code.cpp.rangeanalysis.extensions.ConstantLShiftExprRange
+import experimental.semmle.code.cpp.rangeanalysis.extensions.ConstantRShiftExprRange
+
+Expr getLOp(Operation o) {
```
I can remove these helpers if you want, but I've found them to be useful in a number of cases, so it'd be nice to have them available in some other file so they can be reused.

MathiasVP replied:
I guess the problem this predicate (and `getROp`) solves is that, while `Assignment` is also an operand (and so it has a `getAnOperand` predicate), you often distinguish between the left-hand side or the right-hand side of an `Assignment`. Is that correct?

I'll create an internal issue for this ergonomics problem. For now, I think its best to keep these predicates in this experimental test file.

MathiasVP commented on cpp/ql/lib/experimental/semmle/code/cpp/rangeanalysis/extensions/ConstantLShiftExprRange.qll:
```diff
@@ -0,0 +1,152 @@
+// architecture where the shift value is masked with 0b00011111, but we can't
+// assume the architecture).
+bindingset[val]
+pragma[inline]
```
A `bindingset` annotation causes the predicate to be inlined, so I don't understand why the `inline` pragma is also used? Is it because you really meant to say that both `val` and `l` should be bound already?

gsingh93 replied:
It's been a while since I wrote this code, so I don't exactly remember how it ended up this way, but my guess is I started with `pragma[inline]` and after some refactoring needed the `bindingset`, and I didn't know that `bindingset` results in inlining the predicate. I'll remove the `pragma[inline]` here.

MathiasVP commented on cpp/ql/lib/experimental/semmle/code/cpp/rangeanalysis/extensions/ConstantLShiftExprRange.qll:
```diff
@@ -0,0 +1,152 @@
+        // (a shift of 1) but doing a shift by the upper bound would give 0b01000000.
+        // So if the left shift operation causes an overflow, we just assume the max value
+        // If necessary, we may be able to improve this bound in the future
+        if canLShiftOverflow(lUpper, rUpper, exprMaxVal(this).(int))
```
CodeQL is correct here: `canLShiftOverflow` already requires its third argument to be an `int`, so the explicit `exprMaxVal(this).(int)` cast is redundant.

gsingh93 replied:
Thanks, is there a way to run this linter locally with the vscode plugin or some other way?

MathiasVP commented on cpp/ql/lib/experimental/semmle/code/cpp/rangeanalysis/extensions/ConstantLShiftExprRange.qll:
```diff
@@ -0,0 +1,152 @@
+        // (a shift of 1) but doing a shift by the upper bound would give 0b01000000.
+        // So if the left shift operation causes an overflow, we just assume the max value
+        // If necessary, we may be able to improve this bound in the future
+        if canLShiftOverflow(lUpper, rUpper, exprMaxVal(this).(int))
```
Similarly here.

MathiasVP reviewed (commented):
Here are my first rounds of comments. I think this looks really good!

MathiasVP commented on cpp/ql/lib/experimental/semmle/code/cpp/rangeanalysis/extensions/ConstantLShiftExprRange.qll:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  override predicate dependsOnChild(Expr child) {
+    child = getLeftOperand() or child = getLeftOperand()
```
And the same problem also occurs in the `ConstantRShiftExprRange.qll` file.

gsingh93 commented:
@MathiasVP thanks for the review! I think I've addressed most comments, although instead of merging them into a common base class I just merged them into the same file, mainly due to lack of time on my end.

MathiasVP commented:
> @MathiasVP thanks for the review! I think I've addressed most comments, although instead of merging them into a common base class I just merged them into the same file, mainly due to lack of time on my end.

That sounds good 🙂. The only missing issue I'd like to see fixed in this PR is the copy/paste error in `dependsOnChild` that has two identical disjuncts (i.e., `child = getLeftOperand() or child = getLeftOperand()`), and then I think this PR is good to go!

gsingh93 commented:
Ah, sorry, I missed that in your first round of comments. It should be fixed now.

MathiasVP reviewed (dismissed):
LGTM!

MathiasVP commented:
Actually, let me undo my approval 🙂. CI is currently failing with the following message:
```
File "ql/cpp/ql/lib/experimental/semmle/code/cpp/rangeanalysis/extensions/ConstantShiftExprRange.qll" contains a non-ASCII character at the location marked with `|` in:
2023-01-31T10:00:48.5032066Z ```
2023-01-31T10:00:48.5032493Z f E1
2023-01-31T10:00:48.5033103Z  * has an unsigned type, the value of the result is E1 |
2023-01-31T10:00:48.5033752Z ```
```

I think the issue is the pretty `×` symbol that we unfortunately don't allow in our queries.

gsingh93 commented:
I replaced two unicode characters, hopefully it works now.

MathiasVP reviewed (approved):
LGTM!
